spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala

```diff
@@ -2397,6 +2397,60 @@ trait DeltaProtocolVersionSuiteBase extends QueryTest
     }
   }
 
+  test("Can construct a protocol with column mapping only in writer features") {
+    // Column mapping should appear in both reader and writer features. However,
+    // we allow the construction of invalid protocols such as the one below to allow backward
+    // compatibility. This is to protect against the case a table was created by an older writer
+    // (that somehow serialized the protocol without column mapping in the reader features set)
+    // and then attempted to read the table with a newer reader.
+    Protocol(
+      minReaderVersion = 2,
+      minWriterVersion = 7,
+      readerFeatures = Some(Set.empty),
+      writerFeatures = Some(Set(ColumnMappingTableFeature.name)))
+
+    Protocol(
+      minReaderVersion = 2,
+      minWriterVersion = 7,
+      readerFeatures = Some(Set.empty),
+      writerFeatures = Some(Set(ColumnMappingTableFeature.name, TestWriterFeature.name)))
+
+    // Valid protocol are also allowed.
```

Comment:
Surprised that this is valid. Shouldn't the reader features be missing in this case, because the reader protocol version does not support reader features?

Reply:
Yes, but wouldn't reader version 2 already indicate that this table uses column mapping? And on the other hand reader features are not guaranteed to be understood when the reader version is 2.